crates/bevy_remote/src/lib.rs

```diff
@@ -0,0 +1,849 @@
+pub struct BrpResponse {
+    /// This field is mandatory and must be set to `"2.0"`.
+    pub jsonrpc: &'static str,
+
+    /// The id of the original request.
+    pub id: Option<Value>,
+
+    /// The actual response payload.
+    #[serde(flatten)]
+    pub payload: BrpPayload,
+}
+
+impl BrpResponse {
+    /// Generates a [`BrpResponse`] from an id and a `Result`.
+    #[must_use]
+    pub fn new(id: Option<Value>, result: BrpResult) -> Self {
+        Self {
+            jsonrpc: "2.0",
+            id,
+            payload: BrpPayload::from(result),
+        }
+    }
+}
+
+/// A result/error payload present in every response.
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "snake_case")]
+pub enum BrpPayload {
+    /// `Ok` variant
+    Result(Value),
+    /// `Err` variant
+    Error(BrpError),
+}
+
+impl From<BrpResult> for BrpPayload {
+    fn from(value: BrpResult) -> Self {
+        match value {
+            Ok(v) => Self::Result(v),
+            Err(err) => Self::Error(err),
+        }
+    }
+}
+
+/// An error a request might return.
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct BrpError {
+    /// Defines the general type of the error.
+    pub code: i16,
+    /// Short, human-readable description of the error.
+    pub message: String,
+    /// Optional additional error data.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub data: Option<Value>,
+}
+
+impl BrpError {
+    /// Entity wasn't found.
+    #[must_use]
+    pub fn entity_not_found(entity: Entity) -> Self {
+        Self {
+            code: error_codes::ENTITY_NOT_FOUND,
+            message: format!("Entity {entity} not found"),
+            data: None,
+        }
+    }
+
+    /// Component wasn't found in an entity.
+    #[must_use]
+    pub fn component_not_present(component: &str, entity: Entity) -> Self {
+        Self {
+            code: error_codes::COMPONENT_NOT_PRESENT,
+            message: format!("Component `{component}` not present in Entity {entity}"),
+            data: None,
+        }
+    }
+
+    /// An arbitrary component error. Possibly related to reflection.
+    #[must_use]
+    pub fn component_error(error: anyhow::Error) -> Self {
+        Self {
+            code: error_codes::COMPONENT_ERROR,
+            message: error.to_string(),
+            data: None,
+        }
+    }
+
+    /// An arbitrary internal error.
+    #[must_use]
+    pub fn internal<E: ToString>(error: E) -> Self {
+        Self {
+            code: error_codes::INTERNAL_ERROR,
+            message: error.to_string(),
+            data: None,
+        }
+    }
+
+    /// Attempt to reparent an entity to itself.
+    #[must_use]
+    pub fn self_reparent(entity: Entity) -> Self {
+        Self {
+            code: error_codes::SELF_REPARENT,
+            message: format!("Cannot reparent Entity {entity} to itself"),
+            data: None,
+        }
+    }
+}
+
+/// Error codes used by BRP.
+pub mod error_codes {
```

Comment:
Question: For unfamiliar contributors[^1], would it make sense to add a comment explaining how these integers were chosen?

[^1]: me lol

Reply:
I added a comment about the range reserved by the JSON-RPC specification; outside of that, the Bevy errors numbers are chosen more-or-less arbitrarily.